examples/lang2/dns/src/lib.rs

```diff
@@ -0,0 +1,159 @@
+
+    /// Emitted whenever an address changes.
+    #[ink(event)]
+    struct SetAddress {
+        #[ink(topic)]
+        name: Hash,
+        from: AccountId,
+        #[ink(topic)]
+        old_address: Option<AccountId>,
+        #[ink(topic)]
+        new_address: AccountId,
+    }
+
+    /// Emitted whenver a name is being transferred.
+    #[ink(event)]
+    struct Transfer {
+        #[ink(topic)]
+        name: Hash,
+        from: AccountId,
+        #[ink(topic)]
+        old_owner: Option<AccountId>,
+        #[ink(topic)]
+        new_owner: AccountId,
+    }
+
+    impl DomainNameService {
+        /// Creates a new domain name service contract.
+        #[ink(constructor)]
+        fn new(&mut self) {
+            self.default_address.set(AccountId::from([0x00; 32]));
+        }
+
+        /// Register specific name with caller as owner.
+        #[ink(message)]
+        fn register(&mut self, name: Hash) -> bool {
+            let caller = self.env().caller();
+            if self.name_exists(name) {
+                return false
+            }
+            self.name_to_owner.insert(name, caller);
+            self.env().emit_event(Register { name, from: caller });
+            true
+        }
+
+        /// Set address for specific name.
+        #[ink(message)]
+        fn set_address(&mut self, name: Hash, new_address: AccountId) -> bool {
+            let caller = self.env().caller();
+            let owner = self.get_owner_or_default(name);
+            if caller != owner {
+                return false
+            }
+            let old_address = self.name_to_address.insert(name, new_address);
+            self.env().emit_event(SetAddress {
+                name,
+                from: caller,
+                old_address,
+                new_address,
+            });
+            true
+        }
+
+        /// Transfer owner to another address.
+        #[ink(message)]
+        fn transfer(&mut self, name: Hash, to: AccountId) -> bool {
+            let caller = self.env().caller();
+            let owner = self.get_owner_or_default(name);
+            if caller != owner {
+                return false
+            }
+            let old_owner = self.name_to_owner.insert(name, to);
+            self.env().emit_event(Transfer {
+                name,
+                from: caller,
+                old_owner,
+                new_owner: to,
+            });
+            true
+        }
+
+        /// Get address for specific name.
+        #[ink(message)]
+        fn get_address(&self, name: Hash) -> AccountId {
+            self.get_address_or_default(name)
+        }
+
+        /// Returns `true` if the name already exists.
+        #[ink(message)]
+        fn is_name_exist(&self, name: Hash) -> bool {
+            self.name_exists(name)
+        }
+
+        fn name_exists(&self, name: Hash) -> bool {
+            if self.name_to_owner.get(&name).is_some() {
+                return true
+            }
+            false
```

Comment:
Could be simplified to just `self.name_to_owner.get(&name).is_some()` as the function body (GitHub sadly can't do multi-line suggestions yet).

Reply:
I wish I could undo some commits. :D